stored_requests/caches/memory/cache_test.go

```diff
@@ -0,0 +1,99 @@
+package memory
+
+import (
+	"context"
+	"encoding/json"
+	"strconv"
+	"testing"
+
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/stored_requests"
+	"github.com/prebid/prebid-server/stored_requests/caches/cachestest"
+)
+
+func TestLRURobustness(t *testing.T) {
+	cachestest.AssertCacheRobustness(t, func() stored_requests.Cache {
+		return NewCache(&config.InMemoryCache{
+			RequestCacheSize: 256 * 1024,
+			ImpCacheSize:     256 * 1024,
+			TTL:              -1,
+		})
+	})
+}
+
+func TestUnboundedRobustness(t *testing.T) {
+	cachestest.AssertCacheRobustness(t, func() stored_requests.Cache {
+		return NewCache(&config.InMemoryCache{
+			RequestCacheSize: 0,
+			ImpCacheSize:     0,
+			TTL:              -1,
+		})
+	})
+}
+
+func TestRaceLRUConcurrency(t *testing.T) {
+	cache := NewCache(&config.InMemoryCache{
+		RequestCacheSize: 256 * 1024,
+		ImpCacheSize:     256 * 1024,
+		TTL:              -1,
+	})
+
+	done := make(chan struct{})
+	go writeLots(cache, done, 100)
+	go readLots(cache, done, 100)
+	go invalidateLots(cache, done, 100)
+	for i := 0; i < 3; i++ {
+		<-done
+	}
+}
+
+func TestRaceUnboundedConcurrency(t *testing.T) {
+	cache := NewCache(&config.InMemoryCache{
+		RequestCacheSize: 0,
+		ImpCacheSize:     0,
+		TTL:              -1,
+	})
+
+	done := make(chan struct{})
+	go writeLots(cache, done, 100)
+	go readLots(cache, done, 100)
+	go invalidateLots(cache, done, 100)
+
+	for i := 0; i < 3; i++ {
+		<-done
+	}
+}
+
+func readLots(cache stored_requests.Cache, done chan<- struct{}, numWrites int) {
+	var s struct{}
+	for i := 0; i < numWrites; i++ {
+		cache.Get(context.Background(), sliceForVal(i), sliceForVal(-i))
+	}
+	done <- s
+}
+
+func writeLots(cache stored_requests.Cache, done chan<- struct{}, numWrites int) {
+	var s struct{}
+	for i := 0; i < numWrites; i++ {
+		cache.Save(context.Background(), mapForVal(i), mapForVal(-i))
+	}
+	done <- s
+}
+
+func invalidateLots(cache stored_requests.Cache, done chan<- struct{}, numWrites int) {
+	var s struct{}
+	for i := 0; i < numWrites; i++ {
+		cache.Invalidate(context.Background(), sliceForVal(i), sliceForVal(i))
```

Comment:
read and write you use `i, -i`, why are you using `i, i` here?

Reply:
no good reason... just thought I may as well invalidate half the values to broaden the test scope a bit.

I can see how it'd be confusing though. D'you think I should change it, or add a comment explaining?

Reply:
With imp invalidation taking positive `i`, we will never invalidate an imp we ever set. Not sure if that is supposed to be a feature of the test or not.

It might make sense to shuffle an array of integers of length `numWrites` for each of the three functions, so as to mix reads, writes, and invalidates. If read, write, and invalidate all take the same time to execute, then I would exepect the current code would write `i`, read`i`, and then invalidate `i`. And we would just test that each one of these can be done while the other types of commands are working on other ids. And if reads are faster than writes, we will soon get to the point where we are reading unset data all the time.

Might make sense to shuffle, and have 3 read routines and 3 write routines to cover all possibilities. I don't think multiple invalidates will be needed, as they will be rare and unlikely to run in parallel.

Reply:
these sound like reasonable ideas... done now (if I understood your suggestion properly).

For what it's worth, though... while working on this, I wrote a version with a race condition (using a `map` with no `RWMutex`) and the race detector seemed to pick it up 5 times in a row, with no failures. So we're probably safe.